### PR TITLE
fix!: restrict exported names with __all__

### PR DESCRIPTION
**Pull request by federicobond**
Refs https://github.com/open-feature/python-sdk/issues/214

Note: this is a breaking change. We should do an ecosystem check to make sure most packages are using the correct imports.

### CODE REVIEW

**Comment by federicobond** [comment]
Should `AbstractProvider` be importable from `openfeature.provider` so we can phase out `openfeature.provider.provider`?

File: openfeature/provider/__init__.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
 from .metadata import Metadata
 
+__all__ = ["ProviderStatus", "FeatureProvider", "Metadata"]
```
Comment (gruebel) [inline openfeature/provider/__init__.py:10]: yeah, let's add here `AbstractProvider` and do we want to expose `ProviderRegistry` too or keep it as an internal component?

**Review by gruebel — COMMENTED** [review]
nice work 💪

File: openfeature/provider/__init__.py (hunk 1 above)
Comment (federicobond) [reply openfeature/provider/__init__.py:10]: Yes, I would expose `AbstractProvider` here, but I'd keep `ProviderRegistry` internal unless there is a use case for exposing it.

**Comment by gruebel** [comment]
I think exposing `AbstractProvider` should be done with your other PR #309 then the circular dependency will be gone.

**Comment by federicobond** [comment]
Something weird with the Codecov uploader, @toddbaert @beeme1mr do you know what might be going on?

**Comment by beeme1mr** [comment]
I think this happens when Codecovs attempts to call the GitHub APIs and are throttled. I just ran one of the failed tests to see if that fixed it. If that doesn't help, I will investigate tomorrow.

**Comment by beeme1mr** [comment]
Unfortunately, it looks like a very common issue. 
https://github.com/codecov/codecov-action/issues/598

**Comment by beeme1mr** [comment]
We should probably only run Codecov once after the python specific tests finish.

**Comment by gruebel** [comment]
Yeah, there is no need to upload the results for each Python version, just one is enough. We can add an if condition to only upload for Python 3.11 as it is the default version for dinner other jobs in the same workflow.

**Comment by gruebel** [comment]
ok, seems like we are not the only ones with the codecov issue https://github.com/codecov/codecov-action/issues/1359 seems like we need to upgrade to v4 with API token.

**Comment by beeme1mr** [comment]
> ok, seems like we are not the only ones with the codecov issue [codecov/codecov-action#1359](https://github.com/codecov/codecov-action/issues/1359) seems like we need to upgrade to v4 with API token.

Unfortunately, that has [challenges too](https://github.com/codecov/feedback/issues/112). Perhaps should should consider replacing CodeCov or skipping the upload step for now.

**Comment by gruebel** [comment]
@beeme1mr an alternative could be https://coveralls.io/ thoughts?

**Comment by federicobond** [comment]
I inclined to not rush with the decision and just ignore upload failures in CI temporarily.

**Review by beeme1mr — APPROVED** [review]

**Pull request merged**